src/transport/common/listener.rs

```diff
@@ -100,17 +91,34 @@ impl DialAddresses {
     }
 }
 
-impl WebSocketListener {
-    /// Create new [`WebSocketListener`]
+/// Socket listening to zero or more addresses.
+pub struct SocketListener {
+    /// Listeners.
+    listeners: Vec<TokioTcpListener>,
+    /// The index in the listeners from which the polling is resumed.
+    poll_index: usize,
+}
+
+/// The type of the socket listener.
+#[derive(Clone, Copy, PartialEq, Eq)]
+pub enum SocketListenerType {
+    /// Listener for TCP.
+    Tcp,
+    /// Listener for WebSocket.
+    WebSocket,
+}
+
+impl SocketListener {
+    /// Create new [`SocketListener`]
     pub fn new(
         addresses: Vec<Multiaddr>,
         reuse_port: bool,
+        ty: SocketListenerType,
```

Comment:
nit: may be use the full `type` naming? `ty` personally associates with "thank you", and not with type :D

Reply:
Have removed this from the public api. `type` must be escaped to not be confused by `type X = ..` by the compiler :D